### PR TITLE
post: cool things we learned backend edition

### PR DESCRIPTION
**Pull request by cexbrayat**
First draft, what do you think?

I always had a note about `Failure Analyzer` I don't know if it's worth to include it.

### CODE REVIEW

File: _posts/2019-02-28-cool-things-we-learned-part-1-backend-edition.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+layout: post
+title: Cool things we learned - part 1 - backend edition
+author: cexbrayat
+tags: ["Kotlin", "Spring", "Docker"]
+description: "We completed a project for a customer using Kotlin/Spring Boot and Angular/TypeScript,
+and we tried and learned a few new things we wanted to share"
+---
+
+In the nearly 7 years of our company, we mostly worked on existing applications for our customers.
+But recently, we completed two projects from scratch for two french customers.
```
Comment (clacote) [inline _posts/2019-02-28-cool-things-we-learned-part-1-backend-edition.md]: french -> French

File: _posts/2019-02-28-cool-things-we-learned-part-1-backend-edition.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+We just completed the second one, a small Progressive Web Application for helping citizens to report issues in their city/street (think "Broken lamp" for example).
+We also built the backend part of the application, and a backoffice to allow the local governments (cities, groups of cities, regions...) to see the reported issues and handle them.
+This is not a new concept, but our customer is an organization that promotes open-source in french administrations.
```
Comment (clacote) [inline _posts/2019-02-28-cool-things-we-learned-part-1-backend-edition.md]: french -> French

File: _posts/2019-02-28-cool-things-we-learned-part-1-backend-edition.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+So this application is open-source and will hopefully be used by french citizen,
```
Comment (clacote) [inline _posts/2019-02-28-cool-things-we-learned-part-1-backend-edition.md]: french -> French

File: _posts/2019-02-28-cool-things-we-learned-part-1-backend-edition.md [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+We used Spring Boot as a framework for the backend,
+and the Kotlin integration is great,
+even if it is still a work in progress for some parts (see below).
+
+## Mockk and SpringMockk
+
+We started the project using [Mockito](https://github.com/mockito/mockito),
+mostly as a reflex as we have been using it for a long time in our Java projects.
+
+But we heard more and more talk about an alternative more suited for Kotlin: [MockK](https://mockk.io/).
```
Comment (jnizet) [inline _posts/2019-02-28-cool-things-we-learned-part-1-backend-edition.md]: I'd remove "talk"

File: _posts/2019-02-28-cool-things-we-learned-part-1-backend-edition.md [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+## REST API documentation
+
+When we develop a backend with a REST API, we try to document it as cleanly as possible.
+[Spring REST Docs](https://spring.io/projects/spring-restdocs) is a great help to do so.
+You can write a really nice documentation only using Asciidoc and tests.
+For example, you can check out the documentation of the REST API we wrote for our Angular trainings:
+http://ponyracer.ninja-squad.com/apidoc.
+
+The current DSL is not really thought for Kotlin though, so, you know what?
```
Comment (jnizet) [inline _posts/2019-02-28-cool-things-we-learned-part-1-backend-edition.md]: thought -> designed

File: _posts/2019-02-28-cool-things-we-learned-part-1-backend-edition.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+It's not yet released because it might be integrated directly into Spring REST Docs in the future (at least we hope so).
+Follow this [issue](https://github.com/spring-projects/spring-restdocs/issues/547) if you are interested.
+
+## MailHog
+
+We are not really Docker experts, but we often use it to ease the setup on our machines.
+This time we had a fairly common thing to do in our application: send emails.
+As you probably know, you need an SMTP server somewhere, so that's always a bit cumbersome to setup and test.
+We stumbled upon a very cool Docker image for [MailHog](https://github.com/mailhog/MailHog),
+that simplified quite a bit our life.
```
Comment (jnizet) [inline _posts/2019-02-28-cool-things-we-learned-part-1-backend-edition.md]: our life quite a bit

File: _posts/2019-02-28-cool-things-we-learned-part-1-backend-edition.md [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+This is what part of our `docker-compose.yml` file looks like:
+
+    smtp:
+      image: mailhog/mailhog
+      container_name: smtp
+      ports:
+        - 25:1025
+        - 8025:8025
+
+MailHog gives you a SMTP testing server (that doesn"t really send emails) with a Web UI allowing to check the emails sent (looks like a classic email client inbox, but you see all the emails sent). Very handy!
```
Comment (jnizet) [inline _posts/2019-02-28-cool-things-we-learned-part-1-backend-edition.md]: doesn"t -> doesn't

File: _posts/2019-02-28-cool-things-we-learned-part-1-backend-edition.md [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+Stay tune for part 2 about the frontend!
```
Comment (jnizet) [inline _posts/2019-02-28-cool-things-we-learned-part-1-backend-edition.md]: tune -> tuned

**Pull request merged**